src/opm/output/eclipse/Inplace.cpp

```diff
@@ -0,0 +1,187 @@
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <fmt/format.h>
+#include <exception>
+
+#include <opm/output/eclipse/Inplace.hpp>
+
+namespace Opm {
+
+namespace {
+static const std::string FIELD_NAME = std::string{"FIELD"};
+static const std::size_t FIELD_ID   = 0;
+}
+
+void Inplace::add(const std::string& region, const std::string& tag, std::size_t region_id, double value) {
+    this->tag_values[region][tag][region_id] = value;
+}
+
+void Inplace::add(const std::string& region, Inplace::Phase phase, std::size_t region_id, double value) {
+    this->phase_values[region][phase][region_id] = value;
+}
+
+void Inplace::add(Inplace::Phase phase, double value) {
+    this->add( FIELD_NAME, phase, FIELD_ID, value );
+}
+
+void Inplace::add(const std::string& tag, double value) {
+    this->add( FIELD_NAME, tag, FIELD_ID, value );
+}
+
+double Inplace::get(const std::string& region, const std::string& tag, std::size_t region_id) const {
+    auto region_iter = this->tag_values.find(region);
+    if (region_iter == this->tag_values.end())
+        throw std::logic_error(fmt::format("No such region: {}", region));
+
+    auto tag_iter = region_iter->second.find(tag);
+    if (tag_iter == region_iter->second.end())
+        throw std::logic_error(fmt::format("No such tag: {}:{}", region, tag));
+
+    auto value_iter = tag_iter->second.find(region_id);
+    if (value_iter == tag_iter->second.end())
+        throw std::logic_error(fmt::format("No such region id: {}:{}:{}", region, tag, region_id));
+
+    return value_iter->second;
+}
+
+double Inplace::get(const std::string& region, Inplace::Phase phase, std::size_t region_id) const {
+    auto region_iter = this->phase_values.find(region);
+    if (region_iter == this->phase_values.end())
+        throw std::logic_error(fmt::format("No such region: {}", region));
+
+    auto phase_iter = region_iter->second.find(phase);
+    if (phase_iter == region_iter->second.end())
+        throw std::logic_error(fmt::format("No such phase: {}:{}", region, static_cast<int>(phase)));
+
+    auto value_iter = phase_iter->second.find(region_id);
+    if (value_iter == phase_iter->second.end())
+        throw std::logic_error(fmt::format("No such region id: {}:{}:{}", region, static_cast<int>(phase), region_id));
+
+    return value_iter->second;
+}
+
+double Inplace::get(Inplace::Phase phase) const {
+    return this->get(FIELD_NAME, phase, FIELD_ID);
+}
+
+double Inplace::get(const std::string& tag) const {
+    return this->get(FIELD_NAME, tag, FIELD_ID);
+}
+
+namespace {
+std::size_t region_max(const std::unordered_map<std::size_t, double>& region_map) {
+    std::size_t max_value = 0;
+    for (const auto& [region_id, _] : region_map) {
+        (void)_;
+        max_value = std::max(max_value, region_id);
```

Comment:
Using `std::max` strictly speaking requires that we include `<algorithm>`.

Reply:
OK